File: src/task/src/image_subscriber_syc.py
#!/usr/bin/env python2
#coding:utf-8

import rospy
from std_msgs.msg import String
import cv2
import message_filters
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import os
import time

save_flag = False
frame_rate = 1
save_dir = "/home/bladesaber/Desktop/"

def imageCallback(imgPtr1, imgPtr2):
    global save_flag
    if save_flag:
        bridge = CvBridge()
        image_1 = bridge.imgmsg_to_cv2(imgPtr1,"bgr8")
        image_2 = bridge.imgmsg_to_cv2(imgPtr2, "bgr8")
        cv2.imwrite(os.path.join(save_dir, "t1.jpg"), image_1)
        cv2.imwrite(os.path.join(save_dir, "t2.jpg"), image_2)

def subCallback(msg):
    global save_flag
    if msg.data=='start' and (not save_flag):
        save_flag = True
        print "start take photo"
    elif msg.data=='shutdown' and save_flag:
        save_flag = False
        print "stop take photo"

    rospy.sleep(1.0)

if __name__ == '__main__':
    rospy.init_node('image_subscriber_syc')

    camera_topic_1 = rospy.get_param('~camera_topic_1', 'camera/image/10')
    camera_topic_2 = rospy.get_param('~camera_topic_2', 'camera/image/20')
    control_topic = rospy.get_param('~control_topic', 'control_topic_1')

    image_1_sub = message_filters.Subscriber(camera_topic_1, Image)
    image_2_sub = message_filters.Subscriber(camera_topic_2, Image)

    sync = message_filters.TimeSynchronizer([image_1_sub, image_2_sub], 100)
    sync.registerCallback(imageCallback)

    sub = rospy.Subscriber('control_topic', String, subCallback)

    rospy.spin()

File: src/task/src/image_subscriber_syc.cpp
#include <iostream>
#include <ros/ros.h>
#include <stdio.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/image_encodings.h>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui/highgui.hpp>
#include "std_msgs/String.h"

#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/time_synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <sensor_msgs/Image.h>

using namespace std;

bool save_flag = false;
int frame_rate = 1;
string save_dir = "/home/bladesaber/Desktop/";

void imageCallback(const sensor_msgs::ImageConstPtr& imgPtr1, const sensor_msgs::ImageConstPtr& imgPtr2)
{
  // cv::imshow("view_1", cv_bridge::toCvShare(imgPtr1, "bgr8")->image);
  // cv::imshow("view_2", cv_bridge::toCvShare(imgPtr2, "bgr8")->image);
  // cv::waitKey(30);

  if(save_flag){
    ROS_INFO("take photo");
    cv::Mat img1 = cv_bridge::toCvShare(imgPtr1, "bgr8")->image;
    cv::Mat img2 = cv_bridge::toCvShare(imgPtr2, "bgr8")->image;

    cv::imwrite(save_dir+"t1.jpg", img1);
    cv::imwrite(save_dir+"t2.jpg", img2);
    }
}

void subCallback(const std_msgs::String& submsg){
    // ROS_INFO("%s", submsg.data.c_str());
    string control_flag = submsg.data.c_str();
    if((control_flag=="start") && (!save_flag)){
      ROS_INFO("start take photo");
      save_flag = true;
    }
    else if ((control_flag=="shutdown") && (save_flag))
    {
      ROS_INFO("stop take photo");
      save_flag = false;
    }
}

int main(int argc, char **argv)
{
    cout<<"image_subscriber_syc.cpp start"<<endl;

    ros::init(argc, argv, "image_subscriber_syc");
    ros::NodeHandle nh;
    
    message_filters::Subscriber<sensor_msgs::Image> image_1_sub(nh, "camera/image/1", 1);
    message_filters::Subscriber<sensor_msgs::Image> image_2_sub(nh, "camera/image/2", 1);
    message_filters::TimeSynchronizer<sensor_msgs::Image, sensor_msgs::Image> sync(image_1_sub, image_2_sub, 10);
    sync.registerCallback(boost::bind(&imageCallback, _1, _2));

    ros::Subscriber sub = nh.subscribe("control_topic", 1000, &subCallback);
    ros::Rate rate(frame_rate);

    while(ros::ok()){
      ros::spinOnce();
      rate.sleep();
    }
    // ros::spin();

    return 0;
}